### DISCUSSION TITLE
Input and hooks

### DISCUSSION THREAD

**Discussion by kbachl** [UI5/webcomponents-react#1142] [Q&A] [answered]
Hi,

so we got a load of form like elements like e.g.: Input
( https://sap.github.io/ui5-webcomponents-react/?path=/docs/inputs-input--default-story )

However, how does one make it really useful by binding it to a property using hooks?

e.g.:  How to bind this input to that useState? One cant even use the regular onChange since the events there are useless CustomEvents<any> without any access to value ... with typescript its locked tight

```
const [myValue, setValue] = useState('value')

<Input
        placeholder={'Enter Value'}
        className=""
        onChange={function noRefCheck(){}}
        onInput={function noRefCheck(){}}
        onSubmit={function noRefCheck(){}}
        onSuggestionItemPreview={function noRefCheck(){}}
        onSuggestionItemSelect={function noRefCheck(){}}
        onSuggestionScroll={function noRefCheck(){}}
        slot=""
        style={{}}
        tooltip=""
    />
```

while in normal react is it as easy as this:

```
const [myValue, setValue] = useState('value')
<input type="text" value={myValue} onChange={(e : React.FormEvent<HTMLInputElement>) => setValue(e.currentTarget.value)}/>
<div>Output: {myValue} </div>
```

**Comment by MarcusNotheis** [comment, marked as answer]
Hey @kbachl,

it should be almost the same with our Input component: you can access the current value via `event.target.value`:
```jsx
const MyComponent = () => {
  const [value, setValue] = useState("");

  const onInputChange = useCallback(
    (event) => {
      setValue(event.target.value);
    },
    [setValue]
  );

  return <Input value={value} onInput={onInputChange} />;
};
```
This sample would update the `value` on every keystroke in the input, you can use `onChange` if you want to update it only once after the input looses focus.

You can also omit the `useCallback` if you want to, that's just a performance optimization.

I hope this helps 🙂

**Reply by kbachl** [reply]
Hello @MarcusNotheis ,

thank you for that example. This indeed works. But regarding the CustomEvent  - why is this done this way and not the default React.FormEvent<HTMLInputElement> used? That way it would be much more easy to work with. I see that using useCallback is a way to solve it, but I wonder why its this way.

**Reply by MarcusNotheis** [reply]
You don't really need the `useCallback` here, writing it this way would work as well:
```jsx
const MyComponent = () => {
  const [value, setValue] = useState("");
  return <Input value={value} onInput={(event) => setValue(event.target.value)} />;
};
```

The origin of the CustomEvent is that we are wrapper around the UI5 Web Components which are firing those Custom Events (actually we are rending a `ui5-input` element when using our `Input` component, not a regular `input`). As far as I know, it's best practice to fire custom events from Web Components but I'm not 100% sure to be honest. Maybe @vladitasev can elaborate a bit more on this topic.

**Reply by kbachl** [reply]
Thats the problem, with typescript it wont work as we got a CustomEvent<Any> here:

```
Property 'value' does not exist on type 'EventTarget'.  TS2339
    33 | 
  > 34 |             <Input value={myValue} onInput={(event) => setValue(event.target?.value)}/>
       |                                                                               ^

```

Basically that callback works because it seems typescript cant check deeper into reacts logic, but the latter one is not possible with typescript

**Reply by MarcusNotheis** [reply]
aaah now I get it, thanks for clarifying one more time :) Let me check whether we can tweak our types, I guess we haven't use inline arrows so far so that issues slipped through.

**Reply by kbachl** [reply]
> 
> 
> aaah now I get it, thanks for clarifying one more time :) Let me check whether we can tweak our types, I guess we haven't use inline arrows so far so that issues slipped through.

It would be awesome if you could make that possible... i know, useCallback is faster and better, but in case you just need it once or go fast it is way better to just go for {(event) => doSth(event.target.value)}

**Reply by MarcusNotheis** [reply]
Hey @kbachl,
I just got reminded about this conversation your other discussion - we tried to improve our typings and released them with `^0.16.0`.
Please let me know whether things have improved now and please let us know if there is more to do.

**Reply by kbachl** [reply]
@MarcusNotheis :thumbsup:  thanks for the update, will upgrade to 0.16 and have a look

**Reply by kbachl** [reply]
@MarcusNotheis you might want to check #1576 - it seems #1545 didnt make it into 0.16.0 leading to a partially broken typescript in terms of code completion not working as interface is missing
